### PR TITLE
refactor: nodes consolidation — WelfordAccumulator, selector dedup, module reorganization

### PR DESCRIPTION
**Pull request by nghorbani**
## Summary

Major refactoring of channel selector, mixer, and visualization nodes (ALL-5300). Reorganizes files and classes for clarity, adds a reusable streaming statistics utility, enforces freeze/unfreeze consistency via base class, consolidates visualization nodes, reorganizes pipeline configs, adds plugin system CI, and adds comprehensive tests.

### Scope

- **9 steps completed** across `cuvis-ai` (this PR), `cuvis-ai-core` (PR #19), and `cuvis-ai-adaclip` (PR #8)
- 6 old source files deleted, 5 new files created
- 10 classes renamed (Band→Channel, FalseRGB→descriptive)
- 8 nodes rewired to use shared `WelfordAccumulator`
- 33 YAML configs updated, 11 example scripts updated, 19 doc files revised
- Pipeline configs reorganized into `anomaly/` subdirectories (`adaclip/`, `deep_svdd/`, `rx/`)
- 8 new test files, ~100 new tests (total ~520)
- Plugin runtime smoke CI and AdaCLIP standalone manifest with examples

### Step 1: Reusable WelfordAccumulator utility
- New `cuvis_ai/utils/welford.py` — streaming mean/variance/covariance accumulator (`nn.Module`, float64 internals, non-persistent buffers)
- Replaced 5 inline Welford implementations and 3 concat-then-compute approaches across 8 consumer nodes: `RXGlobal`, `ScoreToLogit`, `LADGlobal`, `_compute_band_correlation_matrix`, `TrainablePCA`, `LearnableChannelMixer`, `SoftChannelSelector`, `ZScoreNormalizerGlobal`

### Step 2: Freeze/unfreeze consistency via `TRAINABLE_BUFFERS`
- `cuvis-ai-core` PR #19: Added `TRAINABLE_BUFFERS` class attribute to `Node` base class with `__init_subclass__` validation
- 5 nodes declare trainable buffers — base class handles buffer↔parameter conversion automatically
- `LearnableChannelMixer` keeps custom override (uses `nn.Conv2d`)

### Step 3: File reorganization + class renames + visualization split
- `band_selection.py` + `selector.py` → `channel_selector.py`
- `concrete_selector.py` absorbed into `channel_mixer.py` (`ConcreteBandSelector` → `ConcreteChannelMixer`)
- `pca.py` → `dimensionality_reduction.py`
- `visualizations.py` split into `anomaly_visualization.py` + `pipeline_visualization.py`
- `drcnn_tensorboard_viz.py` absorbed into `pipeline_visualization.py` (`DRCNNTensorBoardViz` → `PipelineComparisonVisualizer`)
- 9 class renames enforcing selector/mixer distinction
- `__init__.py` updated — all visualization nodes + mixers now exported
- No backward compatibility stubs — clean break

### Step 4: Node device & IO cleanup
- Removed redundant `.to(device)` calls from `adaclip.py`, `anomaly_visualization.py`, `channel_selector.py`
- Kept `.to(device)` where `statistical_initialization()` creates CPU tensors after `pipeline.to(device)`

### Step 5: Test infrastructure
- Pytest markers (`unit`/`integration`/`slow`) on all 30 test files
- 5 expensive fixtures promoted to session scope
- Config consolidated in `pytest.ini`
- 3 new isolation tests: `test_rx_per_batch.py`, `test_binary_decider.py`, `test_data_node.py`
- Removed ~28 print statements, bare try/except → `pytest.raises`

### Step 6-7: Pipeline configs + examples
- All `class_name` paths updated in 13 pipeline + 17 trainrun YAML configs
- All imports updated in 11 example scripts
- Pipeline configs reorganized from flat `configs/pipeline/` into `anomaly/adaclip/`, `anomaly/rx/`, `anomaly/deep_svdd/`
- Normalized filenames to snake_case
- Updated Hydra defaults in 12 trainrun configs

### Step 8: Documentation
- 19 doc files updated with new class names, import paths, YAML examples
- New sections for `WelfordAccumulator` and `TRAINABLE_BUFFERS` in concept docs
- Deep SVDD configs, examples, and docs cleaned up for consistency
- Docstring coverage: 97.9%

### Step 9: Changelog
- Comprehensive `CHANGELOG.md` entries for both repos

### Plugin system CI and AdaCLIP standalone
- Added plugin runtime smoke CI workflow (`plugin-runtime-smoke.yml`)
- Added AdaCLIP standalone plugin manifest (`configs/plugins/adaclip.yaml`) and 6 example scripts
- Added plugin contract, manifest sync, and runtime smoke test files
- CI workflows install `libgl1`/`libglib2.0-0` system dependencies for plugin imports

### Fixes
- Added insufficient-samples guard to `RXGlobal` and `ScoreToLogit`
- Fixed trainrun callback field name and `channel_selector` weights config
- Fixed Werkzeug CVE-2026-27199 by bumping 3.1.5 → 3.1.6
- Fixed AdaCLIP pipeline node names and synced tuning values across 8 pipeline configs

## Breaking changes

- **Module paths**: `band_selection`, `selector`, `concrete_selector`, `pca`, `visualizations`, `drcnn_tensorboard_viz` → `channel_selector`, `channel_mixer`, `dimensionality_reduction`, `anomaly_visualization`, `pipeline_visualization`
- **Class names**: 10 renames (see Step 3)
- **Pipeline config paths**: flat → categorized subdirectories
- No deprecation stubs or re-exports

## Depends on

- cuvis-ai-core PR #19 (`TRAINABLE_BUFFERS` in `Node` base class)
- cuvis-ai-adaclip PR #8 (AdaCLIP pipeline node renames and tuning value sync)

## Test plan

- [x] `ruff check` passes (0 errors)
- [x] `ruff format --check` passes
- [x] `pytest tests/ -m "not slow and not gpu"` — all tests pass
- [x] `mkdocs build` passes
- [x] `interrogate` docstring coverage ≥ 95%
- [x] CI pipeline passes (all 5 jobs)

### CODE REVIEW

File: configs/plugins/registry.yaml [hunk 1]
```diff
@@ -15,8 +15,9 @@ plugins:
 
   adaclip:
     # AdaCLIP-based anomaly detection with advanced band selection
-    repo: "https://github.com/cubert-hyperspectral/cuvis-ai-adaclip.git"
-    tag: "v0.1.1"
+    # repo: "https://github.com/cubert-hyperspectral/cuvis-ai-adaclip.git"
+    # tag: "v0.1.1"
+    path: "D:/code-repos/cuvis-ai-adaclip/cuvis-ai-adaclip-consolidation"
```
Comment (birkholz-cubert) [inline configs/plugins/registry.yaml]: Shouldn't be a absolute path specific to your setup

File: configs/trainrun/deep_svdd__.yaml
Comment (birkholz-cubert) [inline configs/trainrun/deep_svdd__.yaml:1]: This seems outdated since it does not follow the structure of the other trainrun config files, also seems doubled with deep_svdd.yaml?

File: cuvis_ai/anomaly/deep_svdd.py
Comment (birkholz-cubert) [inline cuvis_ai/anomaly/deep_svdd.py:1]: Shouldn't the DeepSvdd implementation also note some kind of TRAINABLE_BUFFERS, or are they defined elsewhere?

File: cuvis_ai/node/channel_mixer.py [hunk 2]
```diff
@@ -1,25 +1,34 @@
...
+from cuvis_ai_schemas.enums import ExecutionStage
 from cuvis_ai_schemas.execution import Context, InputStream
 from cuvis_ai_schemas.pipeline import PortSpec
 from torch import Tensor
 
+from cuvis_ai.utils.welford import WelfordAccumulator
+
 
 class LearnableChannelMixer(Node):
     """Learnable channel mixer for hyperspectral data reduction (DRCNN-style).
```
Comment (birkholz-cubert) [inline cuvis_ai/node/channel_mixer.py:34]: Shouldn't this also have a TRAINABLE_BUFFER class attribute?

**Review by birkholz-cubert — COMMENTED** [review]
So some node define the TRAINABLE_BUFFERS and some do not, despite having them. Are they defined somewhere else?
Or do we skip them for now?

File: configs/plugins/registry.yaml (hunk 1 above)
Comment (nghorbani) [reply configs/plugins/registry.yaml]: Fixed — replaced with relative path `../../cuvis-ai-sam3`.

File: configs/trainrun/deep_svdd__.yaml
Comment (nghorbani) [reply configs/trainrun/deep_svdd__.yaml:1]: Deleted. This was the old expanded format — `deep_svdd.yaml` (Hydra composition) is the canonical one.

File: cuvis_ai/anomaly/deep_svdd.py
Comment (nghorbani) [reply cuvis_ai/anomaly/deep_svdd.py:1]: Added `TRAINABLE_BUFFERS` to both `ZScoreNormalizerGlobal` (`zscore_mean`, `zscore_std`) and `DeepSVDDCenterTracker` (`_tracked_center`). Also added to `MinMaxNormalizer` in `normalization.py` for consistency.

File: cuvis_ai/node/channel_mixer.py (hunk 2 above)
Comment (nghorbani) [reply cuvis_ai/node/channel_mixer.py:34]: LearnableChannelMixer does not use register_buffer() — it uses nn.Conv2d layers whose weights are already nn.Parameter. TRAINABLE_BUFFERS is specifically for promoting buffers to parameters during unfreeze(), so it does not apply here. No change needed.

**Comment by nghorbani** [comment]
Addressed all review comments:

- registry.yaml: replaced absolute path with relative path
- deep_svdd__.yaml: deleted (redundant with Hydra-composition deep_svdd.yaml)
- deep_svdd.py: added TRAINABLE_BUFFERS to ZScoreNormalizerGlobal and DeepSVDDCenterTracker
- normalization.py: added TRAINABLE_BUFFERS to MinMaxNormalizer for consistency
- channel_mixer.py: no change needed (LearnableChannelMixer uses nn.Conv2d with nn.Parameter, not register_buffer)

Re: general question about TRAINABLE_BUFFERS consistency: the mechanism is defined in cuvis-ai-core Node base class via __init_subclass__. It only applies to nodes that use register_buffer() and want those buffers promotable to nn.Parameter during unfreeze(). Nodes using nn.Parameter directly (e.g. via nn.Conv2d) do not need it.

File: cuvis_ai/anomaly/rx_detector.py [hunk 3]
```diff
@@ -202,65 +199,27 @@ def statistical_initialization(self, input_stream: InputStream) -> None:
...
-            self._n, self._mean, self._M2 = m, mean_b, M2_b
-        else:
-            n, tot = self._n, self._n + m
-            delta = mean_b - self._mean
-            new_mean = self._mean + delta * (m / tot)
-            outer = torch.outer(delta, delta) * (n * m / tot)
-            self._n, self._mean, self._M2 = tot, new_mean, self._M2 + M2_b + outer
+        # Adapt accumulator if actual data channels differ from constructor's num_channels
+        # (e.g., upstream SoftChannelSelector preserves all channels instead of reducing)
+        if X.shape[1] != self._welford._n_features:
```
Comment (rajanish4) [inline cuvis_ai/anomaly/rx_detector.py:224]: _n_features is a private attribute? or is it with '__', which would be private considered?

File: examples/adaclip/plugins.yaml [hunk 4]
```diff
@@ -13,9 +13,9 @@
 ## For production deployment from Git repository:
 plugins:
   adaclip:
-    repo: "https://github.com/cubert-hyperspectral/cuvis-ai-adaclip.git"
-    tag: "v0.1.1"  
-    # path: "D:/code-repos/cuvis-ai-adaclip"  # Local path for development"
+    # repo: "https://github.com/cubert-hyperspectral/cuvis-ai-adaclip.git"
+    # tag: "v0.1.1"
+    path: "D:/code-repos/cuvis-ai-adaclip/cuvis-ai-adaclip-consolidation"
```
Comment (rajanish4) [inline examples/adaclip/plugins.yaml]: local path?

**Review by birkholz-cubert — APPROVED** [review]

File: cuvis_ai/anomaly/rx_detector.py (hunk 3 above)
Comment (nghorbani) [reply cuvis_ai/anomaly/rx_detector.py:224]: Single underscore (_n_features) is a Python convention for internal/non-public attributes. Double underscore (__name) triggers name mangling. Here _n_features is intentionally internal to WelfordAccumulator — it is only accessed within the class and by RXGlobal which is tightly coupled to it.

File: examples/adaclip/plugins.yaml (hunk 4 above)
Comment (nghorbani) [reply examples/adaclip/plugins.yaml]: This file is deleted in this PR. The adaclip plugin config now lives in configs/plugins/adaclip.yaml using the repo URL (https://github.com/cubert-hyperspectral/cuvis-ai-adaclip.git, tag v0.1.2).
